File: leadergroup/leader_group/report/cost_center_summary_report/cost_center_summary_report.js
// Copyright (c) 2022, Leader Group and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.provide("leadergroup.cost_center_summary_report");

leadergroup.cost_center_summary_report = {
    "open_general_ledger": function(data) {
        if (!data.name) return;
        frappe.route_options = {
            "company": frappe.query_report.get_filter_value('company'),
            "from_date": frappe.query_report.get_filter_value('from_date'),
            "to_date": frappe.query_report.get_filter_value('to_date'),
            "cost_center": frappe.query_report.get_filter_value('cost_center')
        };
        account = data.name;
        frappe.route_options["account"] = account;
        frappe.set_route("query-report", "General Ledger");
    }
};

frappe.query_reports["Cost Center Summary Report"] = {
	"filters": [
        {
            "fieldname": "company",
            "label": __("Company"),
            "fieldtype": "Link",
            "options": "Company",
            "default": frappe.defaults.get_user_default("Company") || frappe.defaults.get_global_default("Company"),
            "reqd": 1,
            "hidden": 1
        },
		{
			"fieldname":"cost_center",
			"label": __("Cost Center"),
			"fieldtype": "MultiSelectList",
			get_data: function(txt) {
				return frappe.db.get_link_options('Cost Center', txt, {
					company: frappe.query_report.get_filter_value("company")
				});
			}
		},
        {
            "fieldname": "from_date",
            "label": __("From Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.add_months(frappe.datetime.get_today(), -1),
            "reqd": 1
        },
        {
            "fieldname": "to_date",
            "label": __("To Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.get_today(),
            "reqd": 1
        },
        {
            "fieldname": "printed_on",
            "label": __("Printed On"),
            "fieldtype": "Data",
            "default": moment(frappe.datetime.now_datetime()).format("MM-DD-YYYY HH:mm:ss"),
            "read_only": 1
        }
	],
    "formatter": function(value, row, column, data, default_formatter) {
        if (data && column.fieldname=="account_name") {
            value = data.account_name || value;

            column.link_onclick = "leadergroup.cost_center_summary_report.open_general_ledger(" + JSON.stringify(data) + ")";
        }
        value = default_formatter(value, row, column, data);
        return value;
    },
    
    onload: function(report) {
        const views_menu = report.page.add_custom_button_group(__('General Ledger Report'));

        report.page.add_custom_menu_item(views_menu, __("General Ledger"), function() {
            var filters = report.get_values();
            frappe.set_route('query-report', 'General Ledger', filters);
        });
    }
};

File: leadergroup/leader_group/report/cost_center_summary_report/cost_center_summary_report.py
# Copyright (c) 2022, Leader Group and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from erpnext import get_company_currency, get_default_company
from erpnext.accounts.report.financial_statements import get_cost_centers_with_children

def execute(filters=None):
	columns, data = [], []
	columns = get_columns(filters)
	
	conditions, filters = get_conditions(filters)

	data = frappe.db.sql("""select A.name, A.account_name, 
		A.account_number, 
		sum(GL.debit) as debit, sum(GL.credit) as credit
		from `tabGL Entry` GL
		left join `tabAccount` A
			on A.name = GL.account
		where GL.docstatus = 1 and GL.is_cancelled = 0 
			{0} group by GL.account;""".format(conditions),filters, as_dict=1)
	return columns, data

def get_columns(filters):
	columns = []
	columns.extend( 
		[
		{
			"fieldname": "account_number",
			"label": _("Account #"),
			"fieldtype": "Data",
			"width": 120
		}, 
		{
			"fieldname": "account_name",
			"label": _("Account Name"),
			"fieldtype": "Link",
			"options": "Account",
			"width": 200
		},
		{
			"fieldname": "debit",
			"label": _("Debit ({0})").format(get_company_currency(get_default_company())),
			"fieldtype": "Float",
			"width": 120
		},
		{
			"fieldname": "credit",
			"label": _("Credit ({0})").format(get_company_currency(get_default_company())),
			"fieldtype": "Float",
			"width": 120
		}
		]
	)
	return columns


def get_conditions(filters):
	conditions=""
	if filters.get("from_date"):
		conditions += ' and posting_date >= %s'  % frappe.db.escape(filters.get("from_date"), percent=False)
	if filters.get("to_date"):
		conditions += ' and posting_date <= %s' % frappe.db.escape(filters.get("to_date"), percent=False)
	if filters.get("cost_center"):
		conditions += ' and cost_center in %(cost_center)s'
		filters.cost_center = frappe.parse_json(filters.get('cost_center'))
		filters.cost_center = get_cost_centers_with_children(filters.cost_center)
	return conditions, filters